tests/e2e/hashkv_test.go

```diff
@@ -0,0 +1,359 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !cluster_proxy
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"go.etcd.io/etcd/client/pkg/v3/fileutil"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/tests/v3/framework/config"
+	fe2e "go.etcd.io/etcd/tests/v3/framework/e2e"
+)
+
+func TestVerifyHashKVAfterCompact(t *testing.T) {
+	tests := []struct {
+		compactedOnTombstoneRev bool
+		hashKVOnTombstoneRev    bool
+	}{
+		{
+			compactedOnTombstoneRev: true,
+			hashKVOnTombstoneRev:    true,
+		},
+		{
+			compactedOnTombstoneRev: false,
+			hashKVOnTombstoneRev:    true,
+		},
+		{
+			compactedOnTombstoneRev: true,
+			hashKVOnTombstoneRev:    false,
+		},
+		{
+			compactedOnTombstoneRev: false,
+			hashKVOnTombstoneRev:    false,
+		},
+	}
+
+	scenarios := []struct {
+		ClusterVersion fe2e.ClusterVersion
+		OnlyOneKey     bool
+	}{
+		{
+			ClusterVersion: fe2e.CurrentVersion,
+			OnlyOneKey:     true,
+		},
+		{
+			ClusterVersion: fe2e.CurrentVersion,
+			OnlyOneKey:     false,
+		},
+		{
+			ClusterVersion: fe2e.QuorumLastVersion,
+			OnlyOneKey:     true,
+		},
+		{
+			ClusterVersion: fe2e.QuorumLastVersion,
+			OnlyOneKey:     false,
+		},
+	}
+
+	for idx, tt := range tests {
+		for _, scenario := range scenarios {
+			t.Run(fmt.Sprintf("#%d - %s - OnlyOneKey=%v", idx, scenario.ClusterVersion, scenario.OnlyOneKey), func(t *testing.T) {
+				ctx := context.Background()
+
+				clus, cliCfg := newClusterForHashKV(t, scenario.ClusterVersion)
+
+				dataset := newTestDatasetInCluster(t, clus, cliCfg, scenario.OnlyOneKey)
+
+				// ensure we can pick up the right revision
+				require.True(t, len(dataset.tombstones) > 2)
+				require.True(t, dataset.tombstones[0]+3 < dataset.tombstones[1])
+				require.True(t, dataset.tombstones[1]+3 < dataset.tombstones[2])
+
+				compactedOnRev := dataset.tombstones[0]
+				if !tt.compactedOnTombstoneRev {
+					compactedOnRev += 3
+				}
+
+				hashKVOnRev := dataset.tombstones[1]
+				if !tt.hashKVOnTombstoneRev {
+					hashKVOnRev += 3
+				}
+
+				cli, err := fe2e.NewEtcdctl(cliCfg, clus.EndpointsGRPC())
+				require.NoError(t, err)
+
+				t.Logf("COMPACT on rev=%d", compactedOnRev)
+				_, err = cli.Compact(ctx, compactedOnRev, config.CompactOption{Physical: true})
+				require.NoError(t, err)
+
+				t.Logf("HashKV on rev=%d", hashKVOnRev)
+				resp, err := cli.HashKV(ctx, hashKVOnRev)
+				require.NoError(t, err)
+
+				require.Len(t, resp, 3)
+				require.True(t, resp[0].Hash != 0)
+				t.Logf("One Hash value is %d", resp[0].Hash)
+
+				require.Equal(t, resp[0].Hash, resp[1].Hash)
+				require.Equal(t, resp[1].Hash, resp[2].Hash)
+			})
+		}
+	}
+}
+
+type datasetInfo struct {
+	keys map[string]struct{}
```

Comment:
You just populate the data for `keys`, but it isn't used at all.

We can remove `datasetInfo`, and just return `tombstones` and `latestRev` in the functions of populating data.